File: KASIKORN Business-Technology Group/Assignment1/Ex4/Ex4.java
package Assignment1.Ex4;

import java.io.*;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.ThreadLocalRandom;

public class Ex4 {
    private int currentId = 0;
    private final DateTimeFormatter dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
    
    // Pre-defined data
    private final String[] provinceCodes = {"HN", "HCM", "DN", "HP", "CT"};
    private final String[] countryCodes = {"VN", "US", "GB", "JP", "SG"};
    private final String[] bankCodes = {"79601001", "79601002", "79601003"};
    private final String[] positions = {"Manager", "Staff", "Director", "Supervisor", "Specialist"};
    private final String[] companySizes = {"S", "M", "L"};
    
    // Simulated name data
    private final String[] firstNames = {"Minh", "Linh", "Tuan", "Hoa", "Nam", "Mai", "Duc", "Lan", "Thanh", "Hong"};
    private final String[] lastNames = {"Nguyen", "Tran", "Le", "Pham", "Hoang", "Phan", "Vu", "Dang", "Bui", "Do"};
    private final String[] companyTypes = {"Trading", "Technology", "Services", "Manufacturing", "Consulting"};

    private String generateThaiId() {
        currentId++;
        int prefix = ThreadLocalRandom.current().nextInt(1, 10);
        String middle = String.format("%09d", currentId);
        int checkDigit = ThreadLocalRandom.current().nextInt(10);
        return String.format("%d%s%d", prefix, middle, checkDigit);
    }

    private String generatePersonName() {
        return lastNames[ThreadLocalRandom.current().nextInt(lastNames.length)] + " " +
               firstNames[ThreadLocalRandom.current().nextInt(firstNames.length)];
    }

    private String generateCompanyName() {
        String companyType = companyTypes[ThreadLocalRandom.current().nextInt(companyTypes.length)];
        String suffix = new String[]{"Group", "Corporation", "Company", "Enterprise"}[ThreadLocalRandom.current().nextInt(4)];
        return String.format("%s %s %d", companyType, suffix, currentId);
    }

    private Map<String, String> generateCustomerRecord(boolean isCorporate) {
        Map<String, String> record = new LinkedHashMap<>();
        LocalDate today = LocalDate.now();
        
        if (isCorporate) {
            String companyName = generateCompanyName();
            record.put("TransactionDate", today.format(dateFormatter));
            record.put("BRID", "80001");
            record.put("CUSTID", "C" + generateThaiId());
            record.put("BANKCODE", bankCodes[ThreadLocalRandom.current().nextInt(bankCodes.length)]);
            record.put("RESIDENT", String.valueOf(ThreadLocalRandom.current().nextInt(2)));
            record.put("CUST_TYPE", "P01");
            record.put("BIRTHDAY", "");
            record.put("FULL_NAME", companyName);
            record.put("TRADENAME", "Trading As " + companyName);
            record.put("ABBNAME", "CO" + currentId);
            record.put("SEX", "");
            record.put("ADDRESS", String.format("Building %d, Street %d, %s",
                    ThreadLocalRandom.current().nextInt(1, 1000),
                    ThreadLocalRandom.current().nextInt(1, 101),
                    provinceCodes[ThreadLocalRandom.current().nextInt(provinceCodes.length)]));
            record.put("PHONE", String.format("+84%09d", ThreadLocalRandom.current().nextInt(100000000, 1000000000)));
            record.put("EMAIL", String.format("contact%d@%s.com", 
                    currentId, companyName.toLowerCase().replace(" ", "")));
            record.put("PROVINCE_CODE", provinceCodes[ThreadLocalRandom.current().nextInt(provinceCodes.length)]);
            record.put("COUNTRY_CODE", countryCodes[ThreadLocalRandom.current().nextInt(countryCodes.length)]);
            record.put("IDENTITY_NUM", String.format("%010d", ThreadLocalRandom.current().nextInt(1000000000, 2000000000)));
            record.put("IDENTITY_DATE", today.minusDays(ThreadLocalRandom.current().nextInt(365, 3651)).format(dateFormatter));
            record.put("SIZEOFCO", companySizes[ThreadLocalRandom.current().nextInt(companySizes.length)]);
            record.put("TOTAL_EMPLOYEE", String.valueOf(ThreadLocalRandom.current().nextInt(10, 1001)));
            record.put("TOTAL_ASSETS", String.format("%.2f", ThreadLocalRandom.current().nextDouble(1000000, 100000001)));
            record.put("CAPITAL", String.format("%.2f", ThreadLocalRandom.current().nextDouble(100000, 10000001)));
            record.put("ECOTYPE", String.format("IN%02d", ThreadLocalRandom.current().nextInt(1, 100)));
            record.put("OPNDATE", today.minusDays(ThreadLocalRandom.current().nextInt(365, 3651)).format(dateFormatter));
            record.put("INC_DATE", today.minusDays(ThreadLocalRandom.current().nextInt(365, 3651)).format(dateFormatter));
            record.put("TAXCODE", String.format("%010d", ThreadLocalRandom.current().nextInt(1000000000, 2000000000)));
        } else {
            LocalDate birthDate = today.minusDays(ThreadLocalRandom.current().nextInt(20 * 365, 60 * 365 + 1));
            String fullName = generatePersonName();
            
            record.put("TransactionDate", today.format(dateFormatter));
            record.put("BRID", "90001");
            record.put("CUSTID", "I" + generateThaiId());
            record.put("BANKCODE", bankCodes[ThreadLocalRandom.current().nextInt(bankCodes.length)]);
            record.put("RESIDENT", String.valueOf(ThreadLocalRandom.current().nextInt(2)));
            record.put("CUST_TYPE", "P02");
            record.put("BIRTHDAY", birthDate.format(dateFormatter));
            record.put("FULL_NAME", fullName);
            record.put("TRADENAME", "");
            record.put("ABBNAME", "");
            record.put("SEX", String.valueOf(ThreadLocalRandom.current().nextInt(2)));
            record.put("ADDRESS", String.format("House %d, Street %d, %s",
                    ThreadLocalRandom.current().nextInt(1, 1000),
                    ThreadLocalRandom.current().nextInt(1, 101),
                    provinceCodes[ThreadLocalRandom.current().nextInt(provinceCodes.length)]));
            record.put("PHONE", String.format("+84%09d", ThreadLocalRandom.current().nextInt(100000000, 1000000000)));
            record.put("EMAIL", String.format("%s%d@example.com",
                    fullName.toLowerCase().replace(" ", "."),
                    ThreadLocalRandom.current().nextInt(1, 1000)));
            record.put("PROVINCE_CODE", provinceCodes[ThreadLocalRandom.current().nextInt(provinceCodes.length)]);
            record.put("COUNTRY_CODE", countryCodes[ThreadLocalRandom.current().nextInt(countryCodes.length)]);
            record.put("IDENTITY_NUM", generateThaiId());
            record.put("IDENTITY_DATE", today.minusDays(ThreadLocalRandom.current().nextInt(365, 3651)).format(dateFormatter));
            record.put("POSITION", positions[ThreadLocalRandom.current().nextInt(positions.length)]);
            record.put("INCOME", String.format("%.2f", ThreadLocalRandom.current().nextDouble(15000, 150001)));
            record.put("TAXCODE", String.format("%010d", ThreadLocalRandom.current().nextInt(1000000000, 2000000000)));
            record.put("OPNDATE", today.minusDays(ThreadLocalRandom.current().nextInt(365, 3651)).format(dateFormatter));
        }
        
        return record;
    }

    public int generateDataFile(int numRecords, String outputFile) throws IOException {
        int numCorporate = (int) (numRecords * 0.05);
        int numIndividual = numRecords - numCorporate;
        List<Map<String, String>> records = new ArrayList<>();

        // Generate corporate records
        for (int i = 0; i < numCorporate; i++) {
            records.add(generateCustomerRecord(true));
        }

        // Generate individual records
        for (int i = 0; i < numIndividual; i++) {
            records.add(generateCustomerRecord(false));
        }

        // Write to CSV file
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(outputFile))) {
            // Write header
            if (!records.isEmpty()) {
                writer.write(String.join(",", records.get(0).keySet()));
                writer.newLine();
            }

            // Write records
            for (Map<String, String> record : records) {
                writer.write(String.join(",", record.values()));
                writer.newLine();
            }
        }

        return records.size();
    }

    public static void main(String[] args) {
        Ex4 generator = new Ex4();
        int[] sizes = {10000, 100000, 1000000};

        for (int size : sizes) {
            System.out.printf("%nGenerating %d records...%n", size);
            long startTime = System.currentTimeMillis();
            String outputFile = String.format("customer_data_%drecords.csv", size);

            try {
                int recordsGenerated = generator.generateDataFile(size, outputFile);
                long endTime = System.currentTimeMillis();
                double duration = (endTime - startTime) / 1000.0;

                System.out.printf("Generated %d records in %s%n", recordsGenerated, outputFile);
                System.out.printf("Time taken: %.2f seconds%n", duration);
                System.out.printf("Records per second: %.2f%n", recordsGenerated / duration);
            } catch (IOException e) {
                System.err.println("Error generating data file: " + e.getMessage());
            }
        }
    }
}

File: KASIKORN Business-Technology Group/Assignment1/Ex4/Ex4.py
import pandas as pd
import random
from datetime import datetime, timedelta
import numpy as np
import psutil

class CustomerDataGenerator:
    def __init__(self):
        # Simulated master data
        self.current_id = 0
        
        # Pre-defined data
        self.province_codes = ['HN', 'HCM', 'DN', 'HP', 'CT']
        self.country_codes = ['VN', 'US', 'GB', 'JP', 'SG']
        self.bank_codes = ['79601001', '79601002', '79601003']
        self.positions = ['Manager', 'Staff', 'Director', 'Supervisor', 'Specialist']
        self.company_sizes = ['S', 'M', 'L']
        
        # Simulated name data
        self.first_names = ['Minh', 'Linh', 'Tuan', 'Hoa', 'Nam', 'Mai', 'Duc', 'Lan', 'Thanh', 'Hong']
        self.last_names = ['Nguyen', 'Tran', 'Le', 'Pham', 'Hoang', 'Phan', 'Vu', 'Dang', 'Bui', 'Do']
        self.company_types = ['Trading', 'Technology', 'Services', 'Manufacturing', 'Consulting']
        
    def generate_thai_id(self):
        """Generate unique Thai ID number following specific format"""
        self.current_id += 1
        prefix = random.randint(1, 9)
        middle = str(self.current_id).zfill(9)
        # In real implementation, would include checksum calculation
        check_digit = random.randint(0, 9)
        return f"{prefix}{middle}{check_digit}"
    
    def generate_person_name(self):
        """Generate a random Vietnamese person name"""
        return f"{random.choice(self.last_names)} {random.choice(self.first_names)}"
    
    def generate_company_name(self):
        """Generate a random company name"""
        return f"{random.choice(self.company_types)} {random.choice(['Group', 'Corporation', 'Company', 'Enterprise'])} {self.current_id}"
    
    def generate_customer_record(self, is_corporate=False):
        today = datetime.now()
        
        if is_corporate:
            company_name = self.generate_company_name()
            record = {
                'TransactionDate': today.strftime('%Y-%m-%d'),
                'BRID': '80001',
                'CUSTID': f"C{self.generate_thai_id()}",
                'BANKCODE': random.choice(self.bank_codes),
                'RESIDENT': random.choice(['0', '1']),
                'CUST_TYPE': 'P01',
                'BIRTHDAY': None,
                'FULL_NAME': company_name,
                'TRADENAME': f"Trading As {company_name}",
                'ABBNAME': f"CO{self.current_id}",
                'SEX': None,
                'ADDRESS': f"Building {random.randint(1,999)}, Street {random.randint(1,100)}, {random.choice(self.province_codes)}",
                'PHONE': f"+84{random.randint(100000000,999999999)}",
                'EMAIL': f"contact{self.current_id}@{company_name.lower().replace(' ', '')}.com",
                'PROVINCE_CODE': random.choice(self.province_codes),
                'COUNTRY_CODE': random.choice(self.country_codes),
                'IDENTITY_NUM': str(random.randint(1000000000, 9999999999)),
                'IDENTITY_DATE': (today - timedelta(days=random.randint(365, 3650))).strftime('%Y-%m-%d'),
                'SIZEOFCO': random.choice(self.company_sizes),
                'TOTAL_EMPLOYEE': random.randint(10, 1000),
                'TOTAL_ASSETS': random.uniform(1000000, 100000000),
                'CAPITAL': random.uniform(100000, 10000000),
                'ECOTYPE': f"IN{random.randint(1,99):02d}",
                'OPNDATE': (today - timedelta(days=random.randint(365, 3650))).strftime('%Y-%m-%d'),
                'INC_DATE': (today - timedelta(days=random.randint(365, 3650))).strftime('%Y-%m-%d'),
                'TAXCODE': str(random.randint(1000000000, 9999999999))
            }
        else:
            # Generate random birth date for age between 20-60
            birth_date = today - timedelta(days=random.randint(20*365, 60*365))
            full_name = self.generate_person_name()
            
            record = {
                'TransactionDate': today.strftime('%Y-%m-%d'),
                'BRID': '90001',
                'CUSTID': f"I{self.generate_thai_id()}",
                'BANKCODE': random.choice(self.bank_codes),
                'RESIDENT': random.choice(['0', '1']),
                'CUST_TYPE': 'P02',
                'BIRTHDAY': birth_date.strftime('%Y-%m-%d'),
                'FULL_NAME': full_name,
                'TRADENAME': None,
                'ABBNAME': None,
                'SEX': random.choice(['0', '1']),
                'ADDRESS': f"House {random.randint(1,999)}, Street {random.randint(1,100)}, {random.choice(self.province_codes)}",
                'PHONE': f"+84{random.randint(100000000,999999999)}",
                'EMAIL': f"{full_name.lower().replace(' ', '.')}{random.randint(1,999)}@example.com",
                'PROVINCE_CODE': random.choice(self.province_codes),
                'COUNTRY_CODE': random.choice(self.country_codes),
                'IDENTITY_NUM': self.generate_thai_id(),
                'IDENTITY_DATE': (today - timedelta(days=random.randint(365, 3650))).strftime('%Y-%m-%d'),
                'POSITION': random.choice(self.positions),
                'INCOME': random.uniform(15000, 150000),
                'TAXCODE': str(random.randint(1000000000, 9999999999)),
                'OPNDATE': (today - timedelta(days=random.randint(365, 3650))).strftime('%Y-%m-%d')
            }
            
        return record

    def generate_data_file(self, num_records, output_file):
        """Generate customer data file with specified number of records"""
        records = []
        
        # Calculate number of corporate vs individual customers (5:95 ratio)
        num_corporate = int(num_records * 0.05)
        num_individual = num_records - num_corporate
        
        # Generate corporate records
        for _ in range(num_corporate):
            records.append(self.generate_customer_record(is_corporate=True))
            
        # Generate individual records
        for _ in range(num_individual):
            records.append(self.generate_customer_record(is_corporate=False))
            
        # Convert to DataFrame and save
        df = pd.DataFrame(records)
        df.to_csv(output_file, index=False, encoding='utf-8')
        return len(records)

def main():
    # Initialize generator
    generator = CustomerDataGenerator()
    
    # Generate files with different sizes
    sizes = [10000, 100000, 1000000]
    
    for size in sizes:
        start_time = datetime.now()
        output_file = f"customer_data_{size}records.csv"
        
        # RAM usage before starting
        start_memory = psutil.virtual_memory().used / (1024 ** 2)  # Convert to MB
        print(f"\nGenerating {size} records...")
        
        # Track RAM usage during the process
        records_generated = generator.generate_data_file(size, output_file)
        
        end_time = datetime.now()
        end_memory = psutil.virtual_memory().used / (1024 ** 2)  # Convert to MB
        
        duration = (end_time - start_time).total_seconds()
        
        print(f"Generated {records_generated} records in {output_file}")
        print(f"Time taken: {duration:.2f} seconds")
        print(f"Records per second: {records_generated/duration:.2f}")
        
        # Display memory usage results
        print(f"Memory used before: {start_memory:.2f} MB")
        print(f"Memory used after: {end_memory:.2f} MB")
        print(f"Memory increase: {end_memory - start_memory:.2f} MB")

if __name__ == "__main__":
    main()
